From f2539b2ceabd50939fd3fcc4e1c2afd090c36e23 Mon Sep 17 00:00:00 2001
From: Katelyn Donnellan <Katelyn.Donnellan@intel.com>
Date: Fri, 24 Mar 2023 13:32:00 +0000
Subject: [PATCH 1/3] Subject: [PATCH 1/3] vlib: CPU load measurement and CLI

The patch calculates CPU load based on number of ticks ellapsed in
processing packets by main/worker thread.

New CLI command to query CPU load:
`show cpu load`
---
 src/vlib/cli.c          | 33 +++++++++++++++++++++++++++++++++
 src/vlib/global_funcs.h | 16 ++++++++++++++++
 src/vlib/main.c         | 16 ++++++++++++++++
 src/vlib/main.h         |  6 ++++++
 4 files changed, 71 insertions(+)

diff --git a/src/vlib/cli.c b/src/vlib/cli.c
index 2697c0ae0..0267f4e58 100644
--- a/src/vlib/cli.c
+++ b/src/vlib/cli.c
@@ -883,6 +883,39 @@ VLIB_CLI_COMMAND (show_memory_usage_command, static) = {
 };
 /* *INDENT-ON* */
 
+static clib_error_t *
+show_cpu_load (vlib_main_t * vm, unformat_input_t * input,
+               vlib_cli_command_t * cmd)
+{
+ uword i;
+
+  vlib_cli_output (vm, "%10s | %10s | %12s", "Thread", "Core", "Load %");
+
+  for (i = 0; i < vlib_get_n_threads (); i++)
+    {
+      vlib_main_t *vm_i;
+
+      vm_i = vlib_get_main_by_index (i);
+      if (!vm_i)
+        continue;
+
+      vlib_cli_output (vm, "%8u   | %8u   |   %8.2f", i, vm_i->cpu_id,
+              (f64)vm_i->cpu_load_points / 100.0);
+    }
+
+  return 0;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (show_cpu_load_command, static) = {
+  .path = "show cpu load",
+  .short_help = "Show cpu load",
+  .function = show_cpu_load,
+  .is_mp_safe = 1,
+};
+/* *INDENT-ON* */
+
+
 static clib_error_t *
 show_cpu (vlib_main_t * vm, unformat_input_t * input,
 	  vlib_cli_command_t * cmd)
diff --git a/src/vlib/global_funcs.h b/src/vlib/global_funcs.h
index 9dd01fbfb..b8ad35760 100644
--- a/src/vlib/global_funcs.h
+++ b/src/vlib/global_funcs.h
@@ -19,6 +19,22 @@
 #ifndef included_vlib_global_funcs_h_
 #define included_vlib_global_funcs_h_
 
+always_inline u32
+vlib_get_n_threads ()
+{
+  return vec_len (vlib_mains);
+}
+
+always_inline vlib_main_t *
+vlib_get_main_by_index (u32 thread_index)
+{
+  vlib_main_t *vm;
+  vm = vlib_mains[thread_index];
+  ASSERT (vm);
+  return vm;
+}
+
+
 always_inline vlib_main_t *
 vlib_get_main (void)
 {
diff --git a/src/vlib/main.c b/src/vlib/main.c
index bfe97953a..6af6e1ad8 100644
--- a/src/vlib/main.c
+++ b/src/vlib/main.c
@@ -1209,6 +1209,9 @@ dispatch_node (vlib_main_t * vm,
 				      /* n_vectors */ n,
 				      /* n_clocks */ t - last_time_stamp);
 
+  if (n)
+    vm->cpu_load_clocks += t - last_time_stamp;
+
   /* When in interrupt mode and vector rate crosses threshold switch to
      polling mode. */
   if (PREDICT_FALSE ((dispatch_state == VLIB_NODE_STATE_INTERRUPT)
@@ -1938,6 +1941,19 @@ vlib_main_or_worker_loop (vlib_main_t * vm, int is_main)
       /* Record time stamp in case there are no enabled nodes and above
          calls do not update time stamp. */
       cpu_time_now = clib_cpu_time_now ();
+      /* Time to update cpu load? */
+      if (PREDICT_FALSE (cpu_time_now >= vm->cpu_load_interval_end) )
+        {
+          if (vm->cpu_load_interval_start)
+            {
+              vm->cpu_load_points = (vm->cpu_load_clocks * 1e4) /
+                  (cpu_time_now - vm->cpu_load_interval_start);
+            }
+          vm->cpu_load_interval_start = cpu_time_now;
+          vm->cpu_load_interval_end = cpu_time_now + 1e9;
+          vm->cpu_load_clocks = 0;
+        }
+
       vm->loops_this_reporting_interval++;
       now = clib_time_now_internal (&vm->clib_time, cpu_time_now);
       /* Time to update loops_per_second? */
diff --git a/src/vlib/main.h b/src/vlib/main.h
index 45a521a86..985e0c3e3 100644
--- a/src/vlib/main.h
+++ b/src/vlib/main.h
@@ -132,6 +132,12 @@ typedef struct vlib_main_t
   /* Time stamp when main loop was entered (time 0). */
   u64 cpu_time_main_loop_start;
 
+  /* CPU load measurement */
+  u64 cpu_load_interval_start;
+  u64 cpu_load_interval_end;
+  u64 cpu_load_clocks;
+  u32 cpu_load_points;
+
   /* Incremented once for each main loop. */
   volatile u32 main_loop_count;
 
-- 
2.17.1

